sonar-kotlin-plugin/src/main/java/org/sonarsource/kotlin/checks/ImplicitParameterInLambdaCheck.kt

```diff
@@ -0,0 +1,62 @@
+/*
+ * SonarSource Kotlin
+ * Copyright (C) 2018-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonarsource.kotlin.checks
+
+import org.jetbrains.kotlin.psi.KtLambdaArgument
+import org.jetbrains.kotlin.psi.KtLambdaExpression
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.psi.KtTypeReference
+import org.jetbrains.kotlin.psi.KtValueArgument
+import org.sonar.check.Rule
+import org.sonarsource.kotlin.api.AbstractCheck
+import org.sonarsource.kotlin.api.message
+import org.sonarsource.kotlin.plugin.KotlinFileContext
+
+@Rule(key = "S6558")
+class ImplicitParameterInLambdaCheck : AbstractCheck() {
+
+    override fun visitLambdaExpression(lambdaExpression: KtLambdaExpression, kotlinFileContext: KotlinFileContext) {
+        if (lambdaExpression.isArgumentWithImplicitParameterDeclarationWithoutReferenceType()
+            || lambdaExpression.isPropertyWithReferenceTypeAndImplicitParameterDeclaration())
+            kotlinFileContext.reportIssue(psiElement = lambdaExpression, message = buildReportMessage())
+    }
+
+    private fun KtLambdaExpression.isArgumentWithImplicitParameterDeclarationWithoutReferenceType() =
+        isArgument() && hasImplicitParameter() && valueParameters[0].typeReference == null
+
+    private fun KtLambdaExpression.isArgument() =
+        parent is KtLambdaArgument || parent is KtValueArgument
+
+    private fun KtLambdaExpression.isPropertyWithReferenceTypeAndImplicitParameterDeclaration() =
+        isPropertyWithTypeReference() && hasImplicitParameter()
+
+    private fun KtLambdaExpression.hasImplicitParameter() =
+        valueParameters.size == 1 && valueParameters[0].name.equals("it")
+
+    private fun KtLambdaExpression.isPropertyWithTypeReference() =
+        parent is KtProperty && parent.children.filterIsInstance<KtTypeReference>().isNotEmpty()
```

Comment:
Instead of going through all children you can check type reference directly, if you cast parent to KtProperty. You also need to skip parentheses here. So I'd suggest you write something like:

```
(parent.skipParentParentheses() as? KtProperty)?.typeReference != null
```